[DNM] internal executor draft

ZhouXing19 opened the pull request.

ZhouXing19 committed:
sql: introduce query functions under `sql.planner`

Currently, the internal executor always create its own descriptor collections,
txn state, job collection and etc. for its conn executor, even though it's run
underneath a "parent" query. These recreation can unneccesarily reduce the
query efficiency in some use cases, such as when an internal executor is
used under a planner context. In this case, the internal executor is expected
to inherit these info from the planner, rather than creating its own.

To make this rule more explicit, this commit adds a series of query functions
under `sql.planner`. Each of these functions wrap both the init of an internal
executor and the query execution. In this way, the internal executor always
stores the info inherited from the parent planner, and will pass it to its child
conn executor.

fixes https://github.com/cockroachdb/cockroach/issues/69495

Release note: None

ZhouXing19 committed:
sql: refactor adminServer.setUIData with ieProto.WithoutTxn()

This commit provide an example to refactor the current use cases with the new
internal executor interfaces.

In this example, originally, the internal executor was used with a nil txn.
We now replace it with ieProto.WithoutTxn().

Release Note: None

ZhouXing19 committed:
sql: refactor some validation with ieProto.WithTxn() and planner.WithTxn()

This commit is to provides example to refactor the usages of internal executor
with the new interface. Idealy, if a planner is involved, use the query
functions for `sql.planner`. Otherwise, if the query is to run with a not-nil
txn, we should use ieProto.WithTxn().

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82933)
<!-- Reviewable:end -->

ZhouXing19 committed:
sql: introduce sqlutil.InternalExecutorProto and rename SessionBoundInternalExecutorFactory

This is one of the commit series to improve the internal executor usages.

This commit introduce a new struct `sqlutil.InternalExecutorProto`, which
wraps and internal executor factory and a synthetic descriptor collection.
The factory is used to initialize the internal executor.

It's expected that the current existing `InternalExecutor` and
`SessionBoundInternalExecutorFactory` field in structs such as
`sql.ExecutorConfig`, `sql.SchemaChanger` and etc. should all be replaced by
`sqlutil.InternalExecutorProto`. And internal executor should always be
initialized through these the proto.

Also, this commit makes the internal executor factory suit a wider range of
usages -- it can be used to create a general internal executor, not necessarily
a session-bounded one.

Release note: None

ZhouXing19 committed:
sqlutil: introduce InternalExecutorProto.WithTxn() and WithoutTxn()

This commit introduces two functions that allow users to run sql statements
with an internal executor. We intend to limit the usage of a real
internal executor only inside these functions, instead of free-floating
or hanging off certain structs. In other words, we restrict the init of an
internal executor.

The motivation is that if an internal executor is used to run multiple sql
statements in a txn manner, these executions are expected to share info among
their conn executor. While this rule can be easily forgot by the user of
internal executors. Hence we provide an interface that wraps the initialization
of internal executors, so that the users won't need to be worried.

Release note: None
